src/remote.c

```diff
@@ -51,62 +54,66 @@ uint64_t remotehston(const uint32_t max, const char *const str)
 }
 
 #if PC_HOSTED == 0
-static void remote_send_buf(uint8_t *buffer, size_t len)
+/* hex-ify and send a buffer of data */
+static void remote_send_buf(const void *const buffer, const size_t len)
 {
-	uint8_t *p = buffer;
-	char hex[2];
-	do {
-		hexify(hex, (const void *)p++, 1);
-
+	char hex[2] = {};
+	const uint8_t *const data = (const uint8_t *)buffer;
+	for (size_t offset = 0; offset < len; ++offset) {
+		hexify(hex, data + offset, 1U);
 		gdb_if_putchar(hex[0], 0);
 		gdb_if_putchar(hex[1], 0);
-
-	} while (p < (buffer + len));
+	}
 }
 
-static void remote_respond_buf(char respCode, uint8_t *buffer, size_t len)
+/* Send a response with some data following */
+static void remote_respond_buf(const char response_code, const void *const buffer, const size_t len)
 {
 	gdb_if_putchar(REMOTE_RESP, 0);
-	gdb_if_putchar(respCode, 0);
+	gdb_if_putchar(response_code, 0);
 
 	remote_send_buf(buffer, len);
 
 	gdb_if_putchar(REMOTE_EOM, 1);
 }
 
-/* Send response to far end */
-static void remote_respond(char respCode, uint64_t param)
+/* Send a response with a simple result code parameter */
+static void remote_respond(const char response_code, uint64_t param)
 {
-	char buf[35]; /*Response, code, EOM and 2*16 hex nibbles*/
-	char *p = buf;
+	/* Put out the start of response marker and response code */
+	gdb_if_putchar(REMOTE_RESP, false);
+	gdb_if_putchar(response_code, false);
 
-	gdb_if_putchar(REMOTE_RESP, 0);
-	gdb_if_putchar(respCode, 0);
+	/* This defines space for exactly a 64-bit number expressed in hex */
+	char response[16];
+	size_t idx = 0;
 
-	do {
-		*p++ = NTOH(param & 0x0fU);
+	/* Convert the response to hexadecimal */
+	for (; idx < 64 && param; ++idx) {
```

Comment:
This bound is incorrect; it should be 16.

Reply:
Astute observation, thank you!